docs/lazy-adr/adr-005-da-modules-and-components.md

```diff
@@ -0,0 +1,122 @@
+# Lazyledger ADR - Data Availability Module and Components
+
+## Changelog
+
+* 10.06.2021 - Created
+
+## Context
+The core of LazyLedger technology relies on Data Availability schema. For the moment of ADR creation, most DA pieces and
+primitives like `nmt` and `rsmt2d` are already implemented. Furthermore, prototype implementation over the network
+exists as well. However, it still has several unfilled gaps:
+* Slow data retreivability - Validators needs to get data faster and proactively.
+* Fine-grained control over dependencies and 3rd party protocols.
+* Implementation modularity - similar to `nmt` and `rsmt2d`, as it's meant to be used outside of `ll-core` as well.
+* Generic data types and abstractions - contribute to modularity and configurable composition of DA implementations.
+
+The goal of this ADR is to provide a detailed design on how to resolve issues above.
+
+## Detailed Design
+
+### Package
+Data Availability related logic is place in `lazyledger-core/p2p/ipld`. The name and location was good for original
+purpose, as that's a p2p stuff that uses IPLD in it's core. However, currently it makes sense to generalize and rename
+the package to `da`(short for DataAvailability) or even move it in it's own repository, what could be done later with
+ease if package is properly designed.
+
+### Data Types
+Currently, `DAHeader` and block's `Data` are defined in `ll-core/types`. However, with modularity in mind and plan for
+DA to be used outside of `ll-core`, e.g. in `optimint`, it now make sense to move them to new `da` package.
+Also, `Data` contains information unrelated to DA in general like Txs, so it makes sense for DA to rely only on
+`NamespacedShare`, leaving the process of preparing shares of any data to user.
+
+### Interface
+```go
+// DataAvailability defines an interface for Data Availability related operations over the network.
+// The goal of this interface is to become general enough to later abstract over other DA schemas, e.g. Kate comms,
+// along with different networking schemas.
+// NOTE: Put and Get relies on DAHeader which is used to address and verify integrity of data, and its solely a user's
+// responsibility to manage header persistence and transportation between peers interested in data.
+type DataAvailability interface {
+	// Put serializes Data, computes DAHeader and makes it available over the network.
+	Put(context.Context, *types.Block) (*types.DataAvailabilityHeader, error)
+
+	// Get efficiently retrieves block's Data using given DAHeader from the network.
+	Get(context.Context, *types.DataAvailabilityHeader) (*types.Data, error)
```

Comment:
In line with the above comment, it seems like a better abstraction to only use a DA header here, and not deal with blocks or block data. Instead, leave the latter 2 to ll-core.

Reply:
Yep, that's an early draft for the interface I had a few weeks ago. It's final form is still not clear for me